docs/hugo/content/design/ADR-2023-04-Patch-Collections.md

```diff
@@ -0,0 +1,115 @@
+* CON: Horrible to use in code because you must dereference the property to range over it (or really do much of anything with it).
+* CON: Breaking change for any consumers of our object model.
+* CON: Modifies our Spec and Status object definitions for an issue that is specific to ARM
+
+### Option 2: Remove `omitempty` 
+
+Remove omitempty from slice/map properties across all generated types.
+
+``` go
+MyProperty []Type 'json:"myProperty"'. 
+```
+
+This means that the slice or map will always be serialized, even if empty.
+
+* PRO: Behaves exactly like we want from a serialization perspective.
+* PRO: Not a breaking change for consumers of our object model.
+* CON: Empty collections will always be present when an ASO CR is read from the cluster, even if the user did not specify them at all.
+* CON: If we extend this to all properties, all properties will be present when an ASO CR is read from the cluster, even if the user did not specify them at all.
+* CON: Modifies our Spec and Status object definitions for an issue that is specific to ARM
+* QUERY: It is possible that some Azure services treat `myProperty: null` as different than omitting the property entirely, although we do not actually know of any cases where that is true.
+
+### Option 3: Custom JSON Serialization
+
+Use a custom JSON serialization library that supports an `omitnil` equivalent, such as [jettison](https://github.com/wI2L/jettison).
+
+* PRO: Behaves exactly like we want from a serialization perspective.
+* PRO: Not a breaking change for consumers of our object model.
+* CON: Breaks CRD YAML deserialization for the ASO controller (as the [sigs.k8s.io/yaml]( https://github.com/kubernetes-sigs/yaml) library we use works with standard JSON serialization attributes)
+* CON: Also breaks YAML serialization for `asoctl` (for the same reason)
+* CON: Modifies our Spec and Status object definitions for an issue that is specific to ARM
+
+### Option 4: Remove `omitempty` for selected ARM types only
+
+As for Option 2, above, we apply the change only selected ARM Spec types, leaving the core ASO Spec and Status types unmodified.
+
+Selection occurs through our existing `ObjectModelConfiguration` allowing us to explicitly target collections (and other properties) that we know are problematic.
+
+* PRO: Behaves exactly like we want from a serialization perspective.
+* PRO: Limits the scope of the change to only ARM types. These are not directly used by consumers of our object model.
+* PRO: Limits the scope of the change to only properties that we know are problematic, reducing any potential for unintended consequences.
+* PRO: No change to the YAML serialization for ASO Spec and Status types, so no visible change to our users.
+* CON: We can only configure properties we know have this problem. We may miss some, requiring a new release to fix affected users.
+
+## Decision
+
+Adopt Option 4, as it is a minimal change that addresses the issue without impacting on current users.
```

Comment:
Generally agree this is the best medium term option. I think some debate on at exactly what level we do the specification (i.e. resource, type, or property)